### ISSUE TITLE
The scripts in the string passed as the 1st argument to wxWebViewEdge::SetPage are ignored.

### ISSUE THREAD

**Issue by tefbop** [wxWidgets/wxWidgets#22048]
**Describe the bug**
Assume we have a string:
```
wxString s(R"(
    <html>
    <head>
        <script type="application/javascript" src="foo.js"></script>
        <script type="application/javascript">
            do_stuff();
        </script>
    </head>
    </html>
)");
```

If we run `webview->SetPage(s, url);` on wxMSW, `foo.js` will be ignored.

The behavior is not the same as on wxGTK.

**Expected vs observed behaviour**

If we run `webview->SetPage(s, url);` on wxGTK, both `foo.js` and `do_stuff()` will be executed. However, if run the same statement `webview->SetPage(s, url);` on wxMSW, only `do_stuff()` will be executed, and `foo.js` will be ignored.

I've tried calling `ICoreWebView2Settings::put_IsScriptEnabled` before calling `wxWebView::SetPage`, but it doesn't fix the problem.

I've also tried calling `ICoreWebView2::NavigateToString` which is the back-end of `wxWebViewEdge::SetPage`, and the problem is the same.

Calling `ICoreWebView2Settings::put_IsScriptEnabled` and `ICoreWebView2::NavigateToString` does work, but it is only when wxMSW is not used.

**Platform and version information**
 - wxWidgets version you use: 3.1.5
 - wxWidgets port you use: wxMSW
 - OS and its version: Windows 10

**Comment by vadz** [comment]
Sorry, really no idea about this. Perhaps @TcT2k might think of something?

**Comment by tefbop** [comment]
> Sorry, really no idea about this. Perhaps @TcT2k might think of something?

It's a feature of Edge. To switch on or off this feature, call `put_IsScriptEnabled`.

The default behavior of `NavigateToString` (and its front-end `wxWebView::SetPage`) is to ignore <script ... src="foo.js"></script>, but code in <script>...</script> will be executed. It's like some kind of the same-domain policy. Attribute `src` of <script> is able to refer to a .js file in any domain, so it is off by default.

`settings->put_IsScriptEnabled(true)` remove this restriction, but I succeeded in doing so only when coding with pure Windows API, no wxMSW is involved.

By the way, if `settings->put_IsScriptEnabled(no)`, both <script src="foo.js"></script> and <script>code</script> will be banned. Both of them will have no effect.

**Comment by tefbop** [comment]
`put_IsScriptEnabled` is a method of interface `ICoreWebView2Settings` which is a part of WebView2. For details, see
[WebView2 Win32 C++ ICoreWebView2Settings | Microsoft Docs](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2settings?view=webview2-1.0.1072.54).

`NavigateToString` is a method of interface `ICoreWebView2`, which is also a part of WebView2. For details, see [WebView2 Win32 C++ ICoreWebView2 | Microsoft Docs](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2?view=webview2-1.0.1072.54).

**Comment by vadz** [comment]
Sorry, I misunderstood you, I thought that you meant that setting `IsScriptEnabled = true` didn't work when using wx. If it does work, all we need to do is to (optionally) allow enabling them from wxWebView code, which should be pretty straightforward.

**Comment by tefbop** [comment]
> 

I thought it is straightforward too before trying this:
```
#ifdef _WIN32
    ICoreWebView2 *backend = reinterpret_cast<ICoreWebView2 *>(m_webview->GetNativeBackend());
    ICoreWebView2Settings *settings;
    if (FAILED(backend->get_Settings(&settings))) throw std::runtime_error("get_Settings failed");
    if (FAILED(settings->put_IsScriptEnabled(true))) throw std::runtime_error("put_IsScriptEnabled failed");
#endif
    m_webview->SetPage(sHTML, "unused");
```
It doesn't work. Something mysterious is going on behind the scene. Maybe the timing isn't right, but as the above code shows, the pointers `backend` and `settings` are valid. So I assume they are well initialized.

Additionally, if replace the `SetPage` above with `ICoreWebView2::NavigateToString`, the problem is still the same. This problem doesn't happen in non-wxWidgets applications. They, the non-wxWdigets applications, work just fine. I've checked file "wxWidgets-3.1.5/src/msw/webview_edge.cpp", but cannot find anything that looks suspicious.

**Comment by vadz** [comment]
So it looks that by misunderstanding you I did actually understand the problem correctly? How quaint.

Anyhow, unfortunately I still don't know what the problem is. Have you tried retrieving the script enabled setting later? Maybe something else changes it after you set it to `true`?

**Comment by tefbop** [comment]
> 

Yes, I did try retrieving the setting after calling `SetPage`, and the value is true, which means the setting isn't changed after manually calling `settings->put_IsScriptEnabled(true)` and `SetPage`.

I think the problem may be timing. Maybe I call `SetPage` too early. I'm developing a web crawler-like library to get JavaScript-generated HTML code (instead of static HTML content). Because it's a library, the requests are not directly from human. Maybe the requests are sent too early before webview is fully initialized, but it is just a guess.

Anyway, it's almost 3 days I've spent time on this, trying to figure out what is going on in `wxWebViewEdge` but in vain. So, I gave up using `wxWebViewEdge` this morning, and start using `ICoreWebView2` directly. Wish it will eventually work. :)

**Comment by TcT2k** [comment]
I could also imagine it's timing. If you have a look at the `LoadURL` implementation [here](https://github.com/wxWidgets/wxWidgets/blob/3cc4ef5f7727f3d0c0d74739a7bcdb4065fda137/src/msw/webview_edge.cpp#L575). I've added handling to load a URL when the underlying webview is finished loading. Because it's initialized asynchronously.
Similar handling is missing in `SetPage()` [implementation](https://github.com/wxWidgets/wxWidgets/blob/3cc4ef5f7727f3d0c0d74739a7bcdb4065fda137/src/msw/webview_edge.cpp#L941), which would let it fail silently when called to early. This would in turn obviously result in errors similar to what you are seeing.

**Comment by tefbop** [comment]
> I could also imagine it's timing. If you have a look at the `LoadURL` implementation [here](https://github.com/wxWidgets/wxWidgets/blob/3cc4ef5f7727f3d0c0d74739a7bcdb4065fda137/src/msw/webview_edge.cpp#L575). I've added handling to load a URL when the underlying webview is finished loading. Because it's initialized asynchronously. Similar handling is missing in `SetPage()` [implementation](https://github.com/wxWidgets/wxWidgets/blob/3cc4ef5f7727f3d0c0d74739a7bcdb4065fda137/src/msw/webview_edge.cpp#L941), which would let it fail silently when called to early. This would in turn obviously result in errors similar to what you are seeing.

Why not use a nested message loop?

The callback of `CreateCoreWebView2Environment` is called immediately when `CreateCoreWebView2Environment` is called, but the callback of `CreateCoreWebView2Controller` is called NOT when `CreateCoreWebView2Controller` is called but when when Windows API function `DispatchMessage` is called. `DispatchMessage` is a part of message loop.

Currently my code is like:
```
webview_t::webview_t(...)
{
    m_loop = wxTheApp->GetTraits()->CreateEventLoop();
    wxEventLoopActivator activate(m_loop.get());
    CreateCoreWebView2Environment(...); // Call `webview_t::env_OnCreated` directly.
    m_loop->Run(); // Call `webview_t::cntlr_OnCreated`.
}
HRESULT webview_t::env_OnCreated(HRESULT errcode, ICoreWebView2Environment *env)
{
    env->CreateCoreWebView2Controller(..., webview_t::cntlr_OnCreated);
}
HRESULT webview_t::cntlr_OnCreated(HRESULT errcode, ICoreWebView2Controller *cntlr)
{
    cntlr->get_CoreWebView2(&m_backend);
    m_loop->Exit();
}
```
The benefit of using nested message loop is that we make sure when the function returns, we have a valid `ICoreWebView2Controller` and a valid `ICoreWebView2`. Plus, we don't need to yield CPU circles (with `wxYield`?) to the outer message loop. Yielding is wasting CPU circles itself. However, the shortage of using nested message loop is that we might mess up the order of events. For example, in the above code, the webview won't show up if its owner/parent isn't visible when `get_CoreWebView2` is called. Therefore, in my real code, I put a workaround `m_parent->Show(true);` before calling `m_loop->Run()`.

**Comment by TcT2k** [comment]
I see how using a seperate event loop can make the asynchronous initialization more linear. But I don't think it makes much sense to the average `wxWebView` user to have to wait for the initialization during the creation of the control.
I know usually it won't take a long time to initialize but in some cases like when there is no user data already created or on a slow machine it might take some time.

**Comment by tefbop** [comment]
> I see how using a seperate event loop can make the asynchronous initialization more linear. But I don't think it makes much sense to the average `wxWebView` user to have to wait for the initialization during the creation of the control. I know usually it won't take a long time to initialize but in some cases like when there is no user data already created or on a slow machine it might take some time.

I was actually hoping that the idea may be helpful for other `wxWebViewEdge` codes as well. If I remember it correctly, when processing JavaScript code, `wxWebViewEdge` uses `wxYield` to go back to main message loop, because the back-end methods run asynchronously. In this way `wxWebViewEdge` keeps itself synchronous to the caller. In my humble opinion, this design isn't better than a nested message loop, because that `wxYield` strategy is a busy-waiting. When running lengthy JavaScript code, e.g. jQuery, it rockets my CPU usage every time, and I have to shutdown my choked virtual machine. I can imagine that the same problem would happen when it runs on a slow (real) machine.

If `wxWebViewEdge` isn't going to provide asynchronous methods, then it's inevitable that `wxWebViewEdge` has to let its caller wait until the underlying asynchronous operations finish. Letting the caller wait is the nature of a synchronous operation.

Whether using a nested message loop or calling `wxYield` in a busy-waiting loop, other GUI widgets are not supposed to be affected, because the message loop is still running.

**Comment by TcT2k** [comment]
I agree that it could be useful to use a separate message loop in `RunScript()` if possible.
But usually using `RunScriptAsync()` should be preferred as all underlying backends (apart from IE) execute asynchronously anyway.
